corehq/apps/userreports/util.py

```diff
@@ -342,3 +342,12 @@ def wrap_report_config_by_type(config, allow_deleted=False):
         }[doc_type].wrap(config)
     except KeyError:
         raise ReportConfigurationNotFoundError()
+
+
+def get_domain_for_ucr_table_name(table_name):
+    if table_name.startswith(UCR_TABLE_PREFIX):
+        return table_name.split('_')[1]
+    elif table_name.startswith(LEGACY_UCR_TABLE_PREFIX):
+        return table_name.split('_')[2]
+    else:
+        raise ValueError(f"Expected {table_name} to start with {UCR_TABLE_PREFIX} or {LEGACY_UCR_TABLE_PREFIX}")
```

Comment:
This makes assumptions about the number of underscores in `UCR_TABLE_PREFIX` and `LEGACY_UCR_TABLE_PREFIX`, and also assumes that the domain name does not contain any underscores (I'm not sure if that's possible or not). What do you think of this instead, which is safer:
```suggestion
    if ucr_table_name.startswith(UCR_TABLE_PREFIX):
        return ucr_table_name[len(UCR_TABLE_PREFIX):]
    if table_name.startswith(LEGACY_UCR_TABLE_PREFIX):
        return ucr_table_name[len(LEGACY_UCR_TABLE_PREFIX):]
    raise ValueError(f"Expected {table_name} to start with {UCR_TABLE_PREFIX} or {LEGACY_UCR_TABLE_PREFIX}")
```

Reply:
The domain is encoded in https://github.com/dimagi/commcare-hq/blob/cf1944c7cd26b39091fdd3928a99a470357c0fa6/corehq/apps/userreports/util.py#L191 prior to being used in the table name, so underscores in the domain will be changed to hyphens in the table name. That being said, I also don't love the assumption on number of underscores.

The main issue with your proposed change is that it is not just the prefix that needs to be removed. A typical table name looks like `ucr_graham-report-test-up_219138135f994da_9f6cf0b0`, which includes a hash of the table id and domain appended to the end. I could expand upon your solution to also truncate the end of this complete name to extract the domain, but I think at that point the assumption of underscores is a better bet. What do you think?

Reply:
Hopefully we never allow underscores in domain names (currently we do not). That would also break this in a way that I think would make it impossible to confidently extract the domain name.

I think you know this, but just for completeness I'll say it anyway. It's very import that we extract the exact domain name and don't, for example, extract a portion of the name since that could match some other domain for which we do not want to be deleting UCRs.

Reply:
> It's very import that we extract the exact domain name and don't, for example, extract a portion of the name since that could match some other domain for which we do not want to be deleting UCRs.

Yeah I think you should have led with this 😄 while this may have been in the back of my head, having that stated certainly changes the way I look at this, and I'm more inclined to agree with you.

> I see that table_id sometimes contains underscores as well, so the number of underscores is unknown.

I see that table_id is also escaped before being used, so similar to the domain name can we not confidently say it will not contain underscores? Also, even if that is true I think it is irrelevant. Our assumption doesn't hinge on the total number of underscores, just on the number of underscores between the prefix and the domain name (the legacy prefix contains an additional underscore).

That being said, I can see how using the prefix and regex to obtain the domain name is safer, and am not opposed to making that change.

Reply:
> The underscore is not changed by .encode('unicode-escape').decode('utf-8').

Yeah realized this after making this comment. My b.

> it is curious why get_table_name uses unicode-escape at all?

I'm wondering the same thing. By the time the domain name makes it to this point, we are most certainly dealing with a string with only allowed characters for a domain.